{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "82110c4d-257e-43e6-a5fc-1dafd08c5725",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- \u001b[1mAttaching core tidyverse packages\u001b[22m -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- tidyverse 2.0.0 --\n",
      "\u001b[32mv\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.3     \u001b[32mv\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32mv\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32mv\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32mv\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32mv\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32mv\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32mv\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32mv\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "-- \u001b[1mConflicts\u001b[22m -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- tidyverse_conflicts() --\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mi\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Warning message:\n",
      "\"package 'circlize' was built under R version 4.3.3\"\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(circlize)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c43fe461-8629-437e-a041-8821a31899f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_scheme <- list(\n",
    "    # Major category: X\n",
    "    X = \"#A9A9A9\",  # Dark Grey for the category X\n",
    "\n",
    "    # Major category: P\n",
    "    terminal_protein = \"#FFA07A\",      # Light Salmon\n",
    "    spacer = \"#FF8C00\",                # Dark Orange\n",
    "    reverse_transcriptase = \"#FFD700\", # Gold\n",
    "    RNAse_h = \"#FFC107\",               # Orange Red\n",
    "\n",
    "    # Major category: S\n",
    "    pre_s1 = \"#87CEEB\",                # Sky Blue\n",
    "    pre_s2 = \"#4682B4\",                # Steel Blue\n",
    "    surface_antigen = \"#1E90FF\",       # Dodger Blue\n",
    "\n",
    "    # Major category: C\n",
    "    pre_core = \"#98FB98\",              # Pale Green\n",
    "    core = \"#2E8B57\",                  # Sea Green\n",
    "\n",
    "    # Repeated Colors for Major Categories\n",
    "    P = \"#FF8C00\",                     # Dark Orange for category label\n",
    "    S = \"#1E90FF\",                     # Dodger Blue for category label\n",
    "    C = \"#2E8B57\"                      # Sea Green for category label\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "627da9c2-299b-4868-8709-e7dfe216c3f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "loading association files...\n",
      "\n",
      "loading amino acid to wg position files...\n",
      "\n",
      "loading gene location file...\n",
      "\n",
      "loading per site hmp file...\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'per_site_hmp_files' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'per_site_hmp_files' not found\nTraceback:\n",
      "1. map(per_site_hmp_files, readRDS)",
      "2. map_(\"list\", .x, .f, ..., .progress = .progress)",
      "3. vctrs_vec_compat(.x, .purrr_user_env)"
     ]
    }
   ],
   "source": [
    "setwd(\"~/hbv_covar3\")\n",
    "genotypes <- c(\"A\", \"B\", \"C\", \"D\")\n",
    "genes <- c(\"C\", \"P\", \"S\", \"X\")\n",
    "potential_ovp_threshold <- 0\n",
    "\n",
    "# input file\n",
    "#gene_msa_files <- map(genes, function(x) {\n",
    "#    paste0(\"./amino_acid/HBV_gene\", x, \"_AAseqs_noShift_stopToDot_mafft.fasta\")\n",
    "#})\n",
    "#names(gene_msa_files) <- genes\n",
    "\n",
    "assoc_files <- paste0(\"./analysis/hbv\", genotypes, \"_lasso_nofuse_pair_extraInfo_mspval_hmp_domain.tsv\")\n",
    "names(assoc_files) <- genotypes\n",
    "\n",
    "gene_aa_to_wg_pos_files <- paste0(\"./analysis/hbv\", genotypes, \"_gene_aa_to_genome_pos_perSamp.rds\")\n",
    "names(gene_aa_to_wg_pos_files) <- genotypes\n",
    "\n",
    "gene_msa_pos_file <- \"./QC/step4_translate_gene/gene_msa_pos.rds\"\n",
    "\n",
    "# load data\n",
    "#message(\"loading gene msa files...\")\n",
    "#gene_msa <- map(genes, function(x) {\n",
    "#    do.call(rbind, read.fasta(gene_msa_files[[x]]))\n",
    "#})\n",
    "#names(gene_msa) <- genes\n",
    "\n",
    "message(\"loading association files...\")\n",
    "assocs <- map(assoc_files, function(x) {\n",
    "    read.table(x, header = T, stringsAsFactors = F)\n",
    "})\n",
    "names(assocs) <- genotypes\n",
    "\n",
    "message(\"loading amino acid to wg position files...\")\n",
    "gene_aa_to_wg_pos <- map(gene_aa_to_wg_pos_files, readRDS)\n",
    "names(gene_aa_to_wg_pos) <- genotypes\n",
    "\n",
    "message(\"loading gene location file...\")\n",
    "gene_msa_pos <- readRDS(gene_msa_pos_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f7aac78-6e6d-401b-a757-75a56b1ca76f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'genotype' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'genotype' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "domain_pos_file <- \"./analysis/protein_domain_pos.rds\"\n",
    "domain_pos <- readRDS(domain_pos_file)\n",
    "domain_data <- list()\n",
    "cur_y <- 0.3\n",
    "for (gene in names(domain_pos[[genotype]])) {\n",
    "    cur_y <- cur_y + 0.5\n",
    "    for (domain in names(domain_pos[[genotype]][[gene]])) {\n",
    "        start <- domain_pos[[genotype]][[gene]][[domain]][1]\n",
    "        end <- domain_pos[[genotype]][[gene]][[domain]][2]\n",
    "        if (end < start) {\n",
    "            end <- end + 3664 + 20\n",
    "        }\n",
    "        domain_data[[domain]] <- c(domain, cur_y, cur_y + 0.5, start, end, color_scheme[[domain]])\n",
    "    }\n",
    "}\n",
    "domain_data <- data.frame(do.call(rbind, domain_data), stringsAsFactors = F)\n",
    "colnames(domain_data) <- c(\"name\", \"y_bottom\", \"y_top\", \"start\", \"end\", \"color\")\n",
    "domain_data$y_bottom <- as.numeric(domain_data$y_bottom)\n",
    "domain_data$y_top <- as.numeric(domain_data$y_top)\n",
    "domain_data$start <- as.numeric(domain_data$start)\n",
    "domain_data$end <- as.numeric(domain_data$end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1f870d23-68e2-42bb-8d41-3365a110a08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "domain_pos <- readRDS(domain_pos_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8fd6b1a8-f562-4b9d-a76a-3522ce88954c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$B</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$P</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$terminal_protein</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>192</li></ol>\n",
       "</dd>\n",
       "\t<dt>$spacer</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>193</li><li>385</li></ol>\n",
       "</dd>\n",
       "\t<dt>$reverse_transcriptase</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>386</li><li>740</li></ol>\n",
       "</dd>\n",
       "\t<dt>$RNAse_h</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>741</li><li>897</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$S</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_s1</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>148</li></ol>\n",
       "</dd>\n",
       "\t<dt>$pre_s2</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>149</li><li>204</li></ol>\n",
       "</dd>\n",
       "\t<dt>$surface_antigen</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>205</li><li>442</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$C</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>30</li></ol>\n",
       "</dd>\n",
       "\t<dt>$core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>31</li><li>238</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$X</dt>\n",
       "\t\t<dd><strong>$X</strong> = <style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>200</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$A</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$P</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$terminal_protein</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>192</li></ol>\n",
       "</dd>\n",
       "\t<dt>$spacer</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>193</li><li>385</li></ol>\n",
       "</dd>\n",
       "\t<dt>$reverse_transcriptase</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>386</li><li>740</li></ol>\n",
       "</dd>\n",
       "\t<dt>$RNAse_h</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>741</li><li>897</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$S</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_s1</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>148</li></ol>\n",
       "</dd>\n",
       "\t<dt>$pre_s2</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>149</li><li>204</li></ol>\n",
       "</dd>\n",
       "\t<dt>$surface_antigen</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>205</li><li>442</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$C</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>30</li></ol>\n",
       "</dd>\n",
       "\t<dt>$core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>31</li><li>238</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$X</dt>\n",
       "\t\t<dd><strong>$X</strong> = <style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>200</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$C</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$P</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$terminal_protein</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>192</li></ol>\n",
       "</dd>\n",
       "\t<dt>$spacer</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>193</li><li>385</li></ol>\n",
       "</dd>\n",
       "\t<dt>$reverse_transcriptase</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>386</li><li>740</li></ol>\n",
       "</dd>\n",
       "\t<dt>$RNAse_h</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>741</li><li>897</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$S</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_s1</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>148</li></ol>\n",
       "</dd>\n",
       "\t<dt>$pre_s2</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>149</li><li>204</li></ol>\n",
       "</dd>\n",
       "\t<dt>$surface_antigen</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>205</li><li>442</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$C</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>30</li></ol>\n",
       "</dd>\n",
       "\t<dt>$core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>31</li><li>238</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$X</dt>\n",
       "\t\t<dd><strong>$X</strong> = <style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>200</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$D</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$P</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$terminal_protein</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>193</li></ol>\n",
       "</dd>\n",
       "\t<dt>$spacer</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>194</li><li>385</li></ol>\n",
       "</dd>\n",
       "\t<dt>$reverse_transcriptase</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>386</li><li>740</li></ol>\n",
       "</dd>\n",
       "\t<dt>$RNAse_h</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>741</li><li>897</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$S</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_s1</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>148</li></ol>\n",
       "</dd>\n",
       "\t<dt>$pre_s2</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>149</li><li>204</li></ol>\n",
       "</dd>\n",
       "\t<dt>$surface_antigen</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>205</li><li>442</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$C</dt>\n",
       "\t\t<dd><dl>\n",
       "\t<dt>$pre_core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>30</li></ol>\n",
       "</dd>\n",
       "\t<dt>$core</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>31</li><li>238</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "\t<dt>$X</dt>\n",
       "\t\t<dd><strong>$X</strong> = <style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>200</li></ol>\n",
       "</dd>\n",
       "</dl>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$B] \\begin{description}\n",
       "\\item[\\$P] \\begin{description}\n",
       "\\item[\\$terminal\\_protein] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 192\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$spacer] \\begin{enumerate*}\n",
       "\\item 193\n",
       "\\item 385\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$reverse\\_transcriptase] \\begin{enumerate*}\n",
       "\\item 386\n",
       "\\item 740\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$RNAse\\_h] \\begin{enumerate*}\n",
       "\\item 741\n",
       "\\item 897\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$S] \\begin{description}\n",
       "\\item[\\$pre\\_s1] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 148\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pre\\_s2] \\begin{enumerate*}\n",
       "\\item 149\n",
       "\\item 204\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$surface\\_antigen] \\begin{enumerate*}\n",
       "\\item 205\n",
       "\\item 442\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$C] \\begin{description}\n",
       "\\item[\\$pre\\_core] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 30\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$core] \\begin{enumerate*}\n",
       "\\item 31\n",
       "\\item 238\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$X] \\textbf{\\$X} = \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 200\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$A] \\begin{description}\n",
       "\\item[\\$P] \\begin{description}\n",
       "\\item[\\$terminal\\_protein] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 192\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$spacer] \\begin{enumerate*}\n",
       "\\item 193\n",
       "\\item 385\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$reverse\\_transcriptase] \\begin{enumerate*}\n",
       "\\item 386\n",
       "\\item 740\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$RNAse\\_h] \\begin{enumerate*}\n",
       "\\item 741\n",
       "\\item 897\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$S] \\begin{description}\n",
       "\\item[\\$pre\\_s1] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 148\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pre\\_s2] \\begin{enumerate*}\n",
       "\\item 149\n",
       "\\item 204\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$surface\\_antigen] \\begin{enumerate*}\n",
       "\\item 205\n",
       "\\item 442\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$C] \\begin{description}\n",
       "\\item[\\$pre\\_core] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 30\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$core] \\begin{enumerate*}\n",
       "\\item 31\n",
       "\\item 238\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$X] \\textbf{\\$X} = \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 200\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$C] \\begin{description}\n",
       "\\item[\\$P] \\begin{description}\n",
       "\\item[\\$terminal\\_protein] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 192\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$spacer] \\begin{enumerate*}\n",
       "\\item 193\n",
       "\\item 385\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$reverse\\_transcriptase] \\begin{enumerate*}\n",
       "\\item 386\n",
       "\\item 740\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$RNAse\\_h] \\begin{enumerate*}\n",
       "\\item 741\n",
       "\\item 897\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$S] \\begin{description}\n",
       "\\item[\\$pre\\_s1] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 148\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pre\\_s2] \\begin{enumerate*}\n",
       "\\item 149\n",
       "\\item 204\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$surface\\_antigen] \\begin{enumerate*}\n",
       "\\item 205\n",
       "\\item 442\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$C] \\begin{description}\n",
       "\\item[\\$pre\\_core] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 30\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$core] \\begin{enumerate*}\n",
       "\\item 31\n",
       "\\item 238\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$X] \\textbf{\\$X} = \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 200\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$D] \\begin{description}\n",
       "\\item[\\$P] \\begin{description}\n",
       "\\item[\\$terminal\\_protein] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 193\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$spacer] \\begin{enumerate*}\n",
       "\\item 194\n",
       "\\item 385\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$reverse\\_transcriptase] \\begin{enumerate*}\n",
       "\\item 386\n",
       "\\item 740\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$RNAse\\_h] \\begin{enumerate*}\n",
       "\\item 741\n",
       "\\item 897\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$S] \\begin{description}\n",
       "\\item[\\$pre\\_s1] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 148\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$pre\\_s2] \\begin{enumerate*}\n",
       "\\item 149\n",
       "\\item 204\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$surface\\_antigen] \\begin{enumerate*}\n",
       "\\item 205\n",
       "\\item 442\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$C] \\begin{description}\n",
       "\\item[\\$pre\\_core] \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 30\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$core] \\begin{enumerate*}\n",
       "\\item 31\n",
       "\\item 238\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\$X] \\textbf{\\$X} = \\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 200\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$B\n",
       ":   $P\n",
       ":   $terminal_protein\n",
       ":   1. 1\n",
       "2. 192\n",
       "\n",
       "\n",
       "\n",
       "$spacer\n",
       ":   1. 193\n",
       "2. 385\n",
       "\n",
       "\n",
       "\n",
       "$reverse_transcriptase\n",
       ":   1. 386\n",
       "2. 740\n",
       "\n",
       "\n",
       "\n",
       "$RNAse_h\n",
       ":   1. 741\n",
       "2. 897\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$S\n",
       ":   $pre_s1\n",
       ":   1. 1\n",
       "2. 148\n",
       "\n",
       "\n",
       "\n",
       "$pre_s2\n",
       ":   1. 149\n",
       "2. 204\n",
       "\n",
       "\n",
       "\n",
       "$surface_antigen\n",
       ":   1. 205\n",
       "2. 442\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       ":   $pre_core\n",
       ":   1. 1\n",
       "2. 30\n",
       "\n",
       "\n",
       "\n",
       "$core\n",
       ":   1. 31\n",
       "2. 238\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$X\n",
       ":   **$X** = 1. 1\n",
       "2. 200\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$A\n",
       ":   $P\n",
       ":   $terminal_protein\n",
       ":   1. 1\n",
       "2. 192\n",
       "\n",
       "\n",
       "\n",
       "$spacer\n",
       ":   1. 193\n",
       "2. 385\n",
       "\n",
       "\n",
       "\n",
       "$reverse_transcriptase\n",
       ":   1. 386\n",
       "2. 740\n",
       "\n",
       "\n",
       "\n",
       "$RNAse_h\n",
       ":   1. 741\n",
       "2. 897\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$S\n",
       ":   $pre_s1\n",
       ":   1. 1\n",
       "2. 148\n",
       "\n",
       "\n",
       "\n",
       "$pre_s2\n",
       ":   1. 149\n",
       "2. 204\n",
       "\n",
       "\n",
       "\n",
       "$surface_antigen\n",
       ":   1. 205\n",
       "2. 442\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       ":   $pre_core\n",
       ":   1. 1\n",
       "2. 30\n",
       "\n",
       "\n",
       "\n",
       "$core\n",
       ":   1. 31\n",
       "2. 238\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$X\n",
       ":   **$X** = 1. 1\n",
       "2. 200\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       ":   $P\n",
       ":   $terminal_protein\n",
       ":   1. 1\n",
       "2. 192\n",
       "\n",
       "\n",
       "\n",
       "$spacer\n",
       ":   1. 193\n",
       "2. 385\n",
       "\n",
       "\n",
       "\n",
       "$reverse_transcriptase\n",
       ":   1. 386\n",
       "2. 740\n",
       "\n",
       "\n",
       "\n",
       "$RNAse_h\n",
       ":   1. 741\n",
       "2. 897\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$S\n",
       ":   $pre_s1\n",
       ":   1. 1\n",
       "2. 148\n",
       "\n",
       "\n",
       "\n",
       "$pre_s2\n",
       ":   1. 149\n",
       "2. 204\n",
       "\n",
       "\n",
       "\n",
       "$surface_antigen\n",
       ":   1. 205\n",
       "2. 442\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       ":   $pre_core\n",
       ":   1. 1\n",
       "2. 30\n",
       "\n",
       "\n",
       "\n",
       "$core\n",
       ":   1. 31\n",
       "2. 238\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$X\n",
       ":   **$X** = 1. 1\n",
       "2. 200\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$D\n",
       ":   $P\n",
       ":   $terminal_protein\n",
       ":   1. 1\n",
       "2. 193\n",
       "\n",
       "\n",
       "\n",
       "$spacer\n",
       ":   1. 194\n",
       "2. 385\n",
       "\n",
       "\n",
       "\n",
       "$reverse_transcriptase\n",
       ":   1. 386\n",
       "2. 740\n",
       "\n",
       "\n",
       "\n",
       "$RNAse_h\n",
       ":   1. 741\n",
       "2. 897\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$S\n",
       ":   $pre_s1\n",
       ":   1. 1\n",
       "2. 148\n",
       "\n",
       "\n",
       "\n",
       "$pre_s2\n",
       ":   1. 149\n",
       "2. 204\n",
       "\n",
       "\n",
       "\n",
       "$surface_antigen\n",
       ":   1. 205\n",
       "2. 442\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       ":   $pre_core\n",
       ":   1. 1\n",
       "2. 30\n",
       "\n",
       "\n",
       "\n",
       "$core\n",
       ":   1. 31\n",
       "2. 238\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "$X\n",
       ":   **$X** = 1. 1\n",
       "2. 200\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$B\n",
       "$B$P\n",
       "$B$P$terminal_protein\n",
       "[1]   1 192\n",
       "\n",
       "$B$P$spacer\n",
       "[1] 193 385\n",
       "\n",
       "$B$P$reverse_transcriptase\n",
       "[1] 386 740\n",
       "\n",
       "$B$P$RNAse_h\n",
       "[1] 741 897\n",
       "\n",
       "\n",
       "$B$S\n",
       "$B$S$pre_s1\n",
       "[1]   1 148\n",
       "\n",
       "$B$S$pre_s2\n",
       "[1] 149 204\n",
       "\n",
       "$B$S$surface_antigen\n",
       "[1] 205 442\n",
       "\n",
       "\n",
       "$B$C\n",
       "$B$C$pre_core\n",
       "[1]  1 30\n",
       "\n",
       "$B$C$core\n",
       "[1]  31 238\n",
       "\n",
       "\n",
       "$B$X\n",
       "$B$X$X\n",
       "[1]   1 200\n",
       "\n",
       "\n",
       "\n",
       "$A\n",
       "$A$P\n",
       "$A$P$terminal_protein\n",
       "[1]   1 192\n",
       "\n",
       "$A$P$spacer\n",
       "[1] 193 385\n",
       "\n",
       "$A$P$reverse_transcriptase\n",
       "[1] 386 740\n",
       "\n",
       "$A$P$RNAse_h\n",
       "[1] 741 897\n",
       "\n",
       "\n",
       "$A$S\n",
       "$A$S$pre_s1\n",
       "[1]   1 148\n",
       "\n",
       "$A$S$pre_s2\n",
       "[1] 149 204\n",
       "\n",
       "$A$S$surface_antigen\n",
       "[1] 205 442\n",
       "\n",
       "\n",
       "$A$C\n",
       "$A$C$pre_core\n",
       "[1]  1 30\n",
       "\n",
       "$A$C$core\n",
       "[1]  31 238\n",
       "\n",
       "\n",
       "$A$X\n",
       "$A$X$X\n",
       "[1]   1 200\n",
       "\n",
       "\n",
       "\n",
       "$C\n",
       "$C$P\n",
       "$C$P$terminal_protein\n",
       "[1]   1 192\n",
       "\n",
       "$C$P$spacer\n",
       "[1] 193 385\n",
       "\n",
       "$C$P$reverse_transcriptase\n",
       "[1] 386 740\n",
       "\n",
       "$C$P$RNAse_h\n",
       "[1] 741 897\n",
       "\n",
       "\n",
       "$C$S\n",
       "$C$S$pre_s1\n",
       "[1]   1 148\n",
       "\n",
       "$C$S$pre_s2\n",
       "[1] 149 204\n",
       "\n",
       "$C$S$surface_antigen\n",
       "[1] 205 442\n",
       "\n",
       "\n",
       "$C$C\n",
       "$C$C$pre_core\n",
       "[1]  1 30\n",
       "\n",
       "$C$C$core\n",
       "[1]  31 238\n",
       "\n",
       "\n",
       "$C$X\n",
       "$C$X$X\n",
       "[1]   1 200\n",
       "\n",
       "\n",
       "\n",
       "$D\n",
       "$D$P\n",
       "$D$P$terminal_protein\n",
       "[1]   1 193\n",
       "\n",
       "$D$P$spacer\n",
       "[1] 194 385\n",
       "\n",
       "$D$P$reverse_transcriptase\n",
       "[1] 386 740\n",
       "\n",
       "$D$P$RNAse_h\n",
       "[1] 741 897\n",
       "\n",
       "\n",
       "$D$S\n",
       "$D$S$pre_s1\n",
       "[1]   1 148\n",
       "\n",
       "$D$S$pre_s2\n",
       "[1] 149 204\n",
       "\n",
       "$D$S$surface_antigen\n",
       "[1] 205 442\n",
       "\n",
       "\n",
       "$D$C\n",
       "$D$C$pre_core\n",
       "[1]  1 30\n",
       "\n",
       "$D$C$core\n",
       "[1]  31 238\n",
       "\n",
       "\n",
       "$D$X\n",
       "$D$X$X\n",
       "[1]   1 200\n",
       "\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "domain_pos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72c6404f-14f5-4e9b-b4e2-4d56b5666544",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
